File: julia.py
#!/usr/bin/env python3

# To make the cover image:
#   - maxIterations = 256
#   - uncomment the ``if count == maxIterations...'' line and comment
#   the one after it in julia.py
#   - set resolution in julia.gplot to 500x500
#   - uncomment last line of julia.gplot, comment the one before it


#   julia.py: generate points to plot a Julia set with a given parameter
#
#   Copyright (c) 2016 Trey Boehm
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#   GNU General Public License for more detail.
#   <http://www.gnu.org/licenses/>

from euler import frange

# Define important magic numbers
parameter = complex(-.805, .175)    # Complex parameter for the function
maxIterations = 256                 # Limit on iterations to test points
realDomain = [-1.6, 1.6]            # Horizontal (real) limits
imagRange  = [-1.6, 1.6]            # Vertical (imaginary) limits
delta = (realDomain[1]-realDomain[0])/1600  # Step for 1600x1600 image

# Return how many iterations it takes z**2 + c to diverge
def testMembership(z, c):
    count = 0
    for i in range(maxIterations):
        z = z**2 + c
        if abs(z) > 2:
            break
        count += 1
    return count

# Iterate over all points in the complex plane bounded by realDomain and
# imagRange, using delta as the step size. Run points (z = x+iy) through
# testMembership(z, c) and print a tab-separated line formatted as such:
# xValue    <tab>   yValue  <tab>   result of testMembership(z,c)
def juliaPoints(c):
    x = realDomain[0]
    while x <= realDomain[1]:
        y = imagRange[0]
        while y <= imagRange[1]:
            z = complex(x, y)
            count = testMembership(z, c)
            #if count == maxIterations: print("%f\t%f" % (x, y))
            print("%f\t%f\t%d" % (x, y, count))
            y += delta
        x += delta

# Run the above function if this program is executed.
if __name__ == "__main__":
    juliaPoints(parameter)

File: julia.c
/*
 * julia.c: Generate points to plot a Julia Set with given parameters
 *
 * Copyright (c) 2018 Trey Boehm
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more detail.
 * <http://www.gnu.org/licenses/>
 */

#include <stdio.h>
#include <complex.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "parameters.h"

/* Number of iterations for z = zeqn(z)+c to diverge */
int testMembership(double complex z, double complex parameter);
/* The equation we iterate on */
double complex zeqn(double complex z);

int testMembership(double complex z, double complex parameter) {
    int count;
    for (count = 0; count < MAX_ITERATIONS; count++) {
        z = zeqn(z) + parameter;
        if (cabs(z) > 4) {
            return count;
        }
    }
    return count;
}

double complex zeqn(double complex z) {
    //return (1-cpow(z,4)/24)/cpow((z-cpow(z,3)/6),3);
    //return (1-cpow(z,3)/6)/cpow((z-cpow(z,2)/2),2);
    //return (1-z*z*z/6)/((z-z*z/2)*(z-z*z/2));
    return cpow(z,2);
}

int main(int argc, char *argv[]) {
    double complex z;
    double delta = (REMAX-REMIN)/((double)RESOLUTION);
    double re;
    double im;
    double theta;
    //double complex radius = PARAMETER;
    double radius = 0.707;
    double complex cVal;
    int iterCount;

    char gnuplotCMD[1000];
    char mvCMD[100];
    FILE* fp;
    char buffer[50];

    sprintf(gnuplotCMD, "gnuplot -e \"reset; set terminal pngcairo size %d, %d; set cbrange [1:%d]; load 'jet.pal'; set output 'png/julia.png'; set lmargin 0; set rmargin 0; set bmargin 0; set tmargin 0; unset border; unset colorbox; unset xtics; unset ytics; plot 'data.txt' with image notitle\" 2>/dev/null", RESOLUTION, RESOLUTION, MAX_ITERATIONS);

    system("mkdir -p png");

    for (theta = 0; theta < 2*3.14159; theta += THETA_STEP) {
        fp = fopen("data.txt", "w");
        //cVal = radius*cos(theta) + radius*sin(theta)*I;
        cVal = radius*cexp(theta*I);
        for (re = REMIN; re < REMAX; re = re + delta) {
            for (im = IMMIN; im < IMMAX; im = im + delta) {
                z = re + im * I;
                iterCount = testMembership(z, cVal);
                sprintf(buffer, "%.3f\t%.3f\t%d\n", re, im, iterCount);
                fwrite(buffer, strlen(buffer), 1, fp);

            }
        }
        printf("theta = %.2f/6.28\n", theta);
        fclose(fp);

        system(gnuplotCMD);
        system("rm data.txt");
        sprintf(mvCMD, "mv png/julia.png png/julia-%.3f.png", theta);
        system(mvCMD);
    }

    return 0;
}

